### ISSUE TITLE
More GCC 13 issues: was hidden [-Werror=overloaded-virtual=]

### ISSUE THREAD

**Issue by marxin** [libyui/libyui#91]
Same as #80, but at different locations:

https://build.opensuse.org/package/live_build_log/openSUSE:Factory:Staging:Gcc7/libyui:libyui-ncurses-pkg/standard/x86_64

```
[   36s] [ 13%] Building CXX object src/CMakeFiles/libyui-ncurses-pkg.dir/NCPkgFilterClassification.cc.o
[   41s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/build/src/include/yui/ncurses/NCPopupTable.h:33,
[   41s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelectorStart.h:34,
[   41s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelectorStart.cc:36:
[   41s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/build/src/include/yui/ncurses/NCPopup.h:51:18: error: 'virtual bool NCPopup::postAgain()' was hidden [-Werror=overloaded-virtual=]
[   41s]    51 |     virtual bool postAgain()
[   41s]       |                  ^~~~~~~~~
[   41s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelector.h:48,
[   41s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelectorStart.h:36:
[   41s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPkgPopupDeps.h:96:18: note:   by 'virtual bool NCPkgPopupDeps::postAgain(NCPkgSolverAction)'
[   41s]    96 |     virtual bool postAgain( NCPkgSolverAction action );
[   41s]       |                  ^~~~~~~~~
[   41s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/build/src/include/yui/ncurses/NCPopupTable.h:33,
[   41s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelectorStart.h:34,
[   41s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelectorPluginImpl.cc:30:
[   41s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/build/src/include/yui/ncurses/NCPopup.h:51:18: error: 'virtual bool NCPopup::postAgain()' was hidden [-Werror=overloaded-virtual=]
[   41s]    51 |     virtual bool postAgain()
[   41s]       |                  ^~~~~~~~~
[   41s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelector.h:48,
[   41s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPackageSelectorStart.h:36:
[   41s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses-pkg/src/NCPkgPopupDeps.h:96:18: note:   by 'virtual bool NCPkgPopupDeps::postAgain(NCPkgSolverAction)'
[   41s]    96 |     virtual bool postAgain( NCPkgSolverAction action );
[   41s]       |                  ^~~~~~~~~
[   43s] cc1plus: all warnings being treated as errors
[   43s] make[2]: *** [src/CMakeFiles/libyui-ncurses-pkg.dir/build.make:104: src/CMakeFiles/libyui-ncurses-pkg.dir/NCPackageSelectorStart.cc.o] Error 1
[   43s] make[2]: *** Waiting for unfinished jobs....
[   43s] cc1plus: all warnings being treated as errors
```

https://build.opensuse.org/package/live_build_log/openSUSE:Factory:Staging:Gcc7/libyui:libyui-ncurses/standard/x86_64

```
[   18s] [ 12%] Building CXX object src/CMakeFiles/libyui-ncurses.dir/NCtext.cc.o
[   19s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/build/src/include/yui/YMenuBar.h:28,
[   19s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/src/NCMenuBar.h:29,
[   19s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/src/NCWidgetFactory.h:48,
[   19s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/src/NCWidgetFactory.cc:27:
[   19s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/build/src/include/yui/YMenuWidget.h:98:18: error: 'virtual void YMenuWidget::addItem(YItem*)' was hidden [-Werror=overloaded-virtual=]
[   19s]    98 |     virtual void addItem( YItem * item_disown );
[   19s]       |                  ^~~~~~~
[   19s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/src/NCMenuButton.h:30,
[   19s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/src/NCWidgetFactory.h:49:
[   19s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-ncurses/build/src/include/yui/YMenuButton.h:79:17: note:   by 'YMenuItem* YMenuButton::addItem(const std::string&, const std::string&)'
[   19s]    79 |     YMenuItem * addItem( const std::string & label,
[   19s]       |                 ^~~~~~~
```

https://build.opensuse.org/package/live_build_log/openSUSE:Factory:Staging:Gcc7/libyui:libyui-rest-api/standard/x86_64

```
[   22s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/build/src/include/yui/YMenuBar.h:28,
[   22s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YHttpWidgetsActionHandler.cc:31:
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/build/src/include/yui/YMenuWidget.h:98:18: error: 'virtual void YMenuWidget::addItem(YItem*)' was hidden [-Werror=overloaded-virtual=]
[   22s]    98 |     virtual void addItem( YItem * item_disown );
[   22s]       |                  ^~~~~~~
[   22s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YHttpWidgetsActionHandler.cc:32:
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/build/src/include/yui/YMenuButton.h:79:17: note:   by 'YMenuItem* YMenuButton::addItem(const std::string&, const std::string&)'
[   22s]    79 |     YMenuItem * addItem( const std::string & label,
[   22s]       |                 ^~~~~~~
[   22s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YDumbTabActionHandler.h:22,
[   22s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YHttpWidgetsActionHandler.h:40,
[   22s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YHttpServer.h:26,
[   22s]                  from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YHttpVersionHandler.cc:23:
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YWidgetActionHandler.h:85:18: error: 'virtual void YWidgetActionHandler::activate_widget(YMultiSelectionBox*, YItem*)' was hidden [-Werror=overloaded-virtual=]
[   22s]    85 |     virtual void activate_widget( YMultiSelectionBox * widget, YItem * item ) {};
[   22s]       |                  ^~~~~~~~~~~~~~~
[   22s] In file included from /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YHttpWidgetsActionHandler.h:42:
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YTableActionHandler.h:43:18: note:   by 'virtual void YTableActionHandler::activate_widget(YTable*, YItem*)'
[   22s]    43 |     virtual void activate_widget( YTable * widget, YItem * item ) {};
[   22s]       |                  ^~~~~~~~~~~~~~~
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YWidgetActionHandler.h:74:18: error: 'virtual void YWidgetActionHandler::activate_widget(YSelectionBox*)' was hidden [-Werror=overloaded-virtual=]
[   22s]    74 |     virtual void activate_widget( YSelectionBox * widget ) {};
[   22s]       |                  ^~~~~~~~~~~~~~~
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YTableActionHandler.h:43:18: note:   by 'virtual void YTableActionHandler::activate_widget(YTable*, YItem*)'
[   22s]    43 |     virtual void activate_widget( YTable * widget, YItem * item ) {};
[   22s]       |                  ^~~~~~~~~~~~~~~
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YWidgetActionHandler.h:73:18: error: 'virtual void YWidgetActionHandler::activate_widget(YTimeField*)' was hidden [-Werror=overloaded-virtual=]
[   22s]    73 |     virtual void activate_widget( YTimeField * widget ) {};
[   22s]       |                  ^~~~~~~~~~~~~~~
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YTableActionHandler.h:43:18: note:   by 'virtual void YTableActionHandler::activate_widget(YTable*, YItem*)'
[   22s]    43 |     virtual void activate_widget( YTable * widget, YItem * item ) {};
[   22s]       |                  ^~~~~~~~~~~~~~~
[   22s] /home/abuild/rpmbuild/BUILD/libyui-4.4.8/libyui-rest-api/src/YWidgetActionHandler.h:72:18: error: 'virtual void YWidgetActionHandler::activate_widget(YRadioButton*)' was hidden [-Werror=overloaded-virtual=]
[   22s]    72 |     virtual void activate_widget( YRadioButton * widget ) {};
...
```

**Comment by marxin** [comment]
May I please ping this, @shundhammer ?

**Comment by shundhammer** [comment]
It's on the radar. Sorry for the delay.

**Comment by shundhammer** [comment]
I really question the wisdom of making the intended use of virtual methods practically impossible with this. Seriously, **WHY**?

I'll switch that bogus warning off in libyui-ncurses* and libyui*rest-api. Still, IMHO this is perverting the ideas behind C++. Who came up with that grand idea? This will lead to a giant backlash for GCC.

**Comment by marxin** [comment]
@jwakely: Can you please chime in and explain the mentioned warning?

**Comment by jwakely** [comment]
> I really question the wisdom of making the intended use of virtual methods practically impossible with this. Seriously, **WHY**?

It's not "practically impossible", that's total nonsense.

The warning tells you that the derived class has an overload that hides a virtual function.

```c++
struct Base {
  virtual void foo() { }
};
struct Derived : Base {
  virtual void foo(int) { }
};
int main() {
  Derived d;
  d.foo(); // error: Base::foo() is hidden by overload
}
```

Why would you want to hide the base function like this? It seems undesirable.

The solution is very simple, and makes the code more expressive and clear:

```c++
struct Derived : Base {
  using Base::foo;
  virtual void foo(int) { }
};
```

Now you can call `d.foo()` or `d.foo(1)` because both are visible in the derived class.

If you **really* want to hide the base function in the derived class for some strange reason, you can still use a using-declaration in the derived class but make it private there:


```c++
struct Derived : Base {
  virtual void foo(int) { }
private:
  using Base::foo;
};
```

This suppresses the warning, because now the base `foo()` is not hidden, but it's also not usable on an object with static type `Derived`.

> I'll switch that bogus warning off in libyui-ncurses* and libyui*rest-api. Still, IMHO this is perverting the ideas behind C++. Who came up with that grand idea? This will lead to a giant backlash for GCC, and it will be well-deserved.

There's no need to be so melodramatic. Why do you think the ability to hide virtual functions from the base class is "the ideas behind C++"? That's nonsense as well.

This isn't a bogus warning, it's a useful one that catches accidental mistakes where the signature of a function that is supposed to be a virtual override doesn't match the base class, and actually overloads and hides the base function instead of overriding it.

**Comment by shundhammer** [comment]
Just look at the real-life examples that we already had in libyui.

There are several very concrete examples where it makes perfect sense to add an `addItem()` function with a different signature, even though that may, according to your new definition, "shadow" a base class's `addItem()` function; only now with not just a generic `Item` pointer or reference, but with a very generic `std::string` argument (or several ones) from which a derived `Item` class is created.

That was and is needed in this library, and we won't make our API less useable because of compiler limitations. In particular, we are not going to rename an API function from its natural name to something artificial; very much like we had to do in the bad old C days.

**Comment by jwakely** [comment]
Oh wait, I see **you** suggested renaming it in #84 but that's because you don't understand the warning.

> That over-zealous warning really aims at forcing the application developer to rename one of the methods;

That's nonsense.

**Comment by shundhammer** [comment]
See also https://github.com/libyui/libyui/issues/80#issuecomment-1249658742

But whatever, this is already taking too much time. In parallel, I am also preparing a PR to make this build.

**Comment by mvidner** [comment]
In the end we solved it in #92 without disabling the warnings (see #93 which is part of 92)
Thanks for your comments!

**Comment by pvgoran** [comment]
I'm very sorry to hijack the thread (and to necro-post, on top of it), but it (the thread) was the first - and the best - source with relevant discussion that I found while investigating the similar issue that I encountered.

In a nutshell, I'm trying to compile ArangoDB 3.11 with GCC 14+, and one of the problems that I get is the `overloaded-virtual` warning, which is turned into error by project's build settings. The problem occurs with the `LogContext::Visitor` => `LogContext::OverloadVisitor` => `LogContext::OverloadVisitor` hierarchy from [lib/Logger/LogContext.h](https://github.com/arangodb/arangodb/blob/3.11/lib/Logger/LogContext.h). I tried to fix the code rather than disable the warning, but failed because the recommended approach "add the `using` statement to the derived class" does not work: function hiding is not only intentional here, it's obligatory.

The combination of virtual function in the base class and same-named non-virtual template function in the derived class is used in this class hierarchy to allow a single template function (a lambda, in fact) to serve as compile-time-polymorphic implementation of several virtual functions with different parameter types - a clever but confusing trick. If the functions from the base class are imported into the derived class with `using`, this trick stops working. I suppose this can be rewritten by using a different name for the non-virtual template function from the derived class to prevent name clash and get rid of the warning; however, it's a bit too intrusive for my case.

I wrote a small example program that reproduces the class hierarchy and its usage pattern: https://gist.github.com/pvgoran/68ffac213ec39fa175ed88c0feccce3a. It includes the (commented) attempt to silence the warning with `using` - the attempt that does *not* work.

Additionally, here is the paper that suggests a language change that, among other things, would allow to mark functions as intentionally hiding base class members: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2009/n2928.htm
